### ISSUE TITLE
Define behavior in the presence of TLS Early Data

### ISSUE THREAD

**Issue by notriddle** [whatwg/fetch#888]
https://tools.ietf.org/html/rfc8470 specifies the 425 Response Code, but it is silent about how it should be exposed within the Web APIs.

I presume that it should be defined in such a way that:

* Page resources (destination != "") and navigation will be requested with early data as long as they don't have query parameters and use idempotent methods.

* Script requests are off-by-default, even if they're idempotent, but can explicitly turn it on.

* Subresource integrity will turn it on.

* 425 Too Early does not get reported to script. Like 304 Not Modified.

**Comment by annevk** [comment]
cc @mnot @martinthomson

**Comment by martinthomson** [comment]
> Page resources (destination != "") and navigation will be requested with early data as long as they don't have query parameters and use idempotent methods.

This is something that I think browsers should be able to determine on their own.  The spec can say "if the user agent sends the request in early data, ..." as necessary.

> Script requests are off-by-default, even if they're idempotent, but can explicitly turn it on.

This is a good idea.  Adding `earlyData: bool` to the options struct is fine.

> Subresource integrity will turn it on.

See above regarding browser policy.  SRI is not a good signal here, but I don't think that we need to litigate the point.

> 425 Too Early does not get reported to script. Like 304 Not Modified.

The idea being that the browser retries?  That might not be feasible for requests with streamed bodies.  That would need some work.

**Comment by notriddle** [comment]
> This is something that I think browsers should be able to determine on their own. The spec can say "if the user agent sends the request in early data, ..." as necessary.

Yeah, in retrospect, I probably phrased it wrong. The spec shouldn't be prescribing when a browser will use early data. The spec should be prescribing when the browser must not use early data, allowing but not requiring it to be used in any other case.

* Fetch requests that were fired by scripts `destination == ""` MUST NOT use early data unless the script specifically enables it, in which case they MAY but are not require dot do so.
* Fetch requests with non-empty destination (that is, ones that were fired by forms, links, or page subresources) MAY use TLS Early Data as long as they are GET requests.

> The idea being that the browser retries?

Yeah, exactly.

> That might not be feasible for requests with streamed bodies. That would need some work.

My immediate instinct was that this shouldn't be a problem, since the size of an early data payload should always be a bit smaller than [TCP MSS](https://en.wikipedia.org/wiki/Maximum_segment_size) anyway (otherwise, it wouldn't be zero-round-trip, because it would have to go through a TCP ACK cycle). But that's not actually helpful. The server isn't guaranteed to fire its 425 Too Early response immediately upon receiving the early data payload (if it was able to, it would have ["ignored the extension and and returned a regular 1RTT response"](https://tools.ietf.org/html/rfc8446#section-4.2.10) instead). A browser that sends early data needs to be prepared to recreate the entire request.

So the best answer is probably not using TLS Early Data if the request has a streaming body. Having one little round trip off of a large data transfer doesn't seem worth it anyway.

**Comment by mnot** [comment]
If avoiding exposing 425 can be managed, that would be good, I think.

**Comment by annevk** [comment]
We always buffer the request body (needed for how we do redirects and potentially also 403), except when the request body is an explicit stream in which case we fail for those cases.

It seems to me that we'd always retry when seeing a 425 based on this, except when there was an explicit request body stream. In that case we'd have to pass it on to the caller? Potentially, we should also avoid making early data requests in this case and making that opt-in.

It's unclear how we should handle erroneous scenarios from the RFC. I guess retry is the way to go, and passing on the 425 when we can't retry due to a request body stream.

**Comment by mnot** [comment]
I think you're on the right track.

**Comment by annevk** [comment]
@yutakahirano @yoichio it seems 425 has interaction with streams as well.

**Comment by yoichio** [comment]
As chrome doesn't read response code until whole body is sent, prohibiting stream body retry and disclosing the code is reasonable.

**Comment by annevk** [comment]
Yeah, I guess it should end up being a fairly similar patch to 421. Perhaps they can even share logic.

**Comment by yoichio** [comment]
I tried with https://github.com/web-platform-tests/wpt/pull/28020:
- Chrome retries the request only if request method is `GET`.
- Firefox don't retry the request.

As https://tools.ietf.org/html/rfc8470#section-4, clients don't sending POST during 0RTT.
> Absent other information, clients MAY send requests with safe HTTP methods ([RFC7231], Section 4.2.1) in early data
> when it is available and MUST NOT send unsafe methods (or methods whose safety is not known) in early data.

Thus we should update fetch spec but it is not upload-streaming-blocking?

**Comment by annevk** [comment]
> The server cannot assume that a client is able to retry a request unless the request is received in early data or the Early-Data header field is set to "1".

And elsewhere there is text that this header is not to be set by user agents. (and I'm not sure that setting it in `fetch()` would really test anything in the user agent).

I think that means that unless we can test TLS early data on WPT (do you know?) Firefox's behavior of always passing on the 425 makes the most sense.

And yeah, I'm okay with your tests being sufficient here and leaving further cleanup for later. (I'll repeat my wish for IETF taking more ownership in end-to-end integration of new features.)
